dependencies/aiven-kafka-auth/src/main/java/io/aiven/kafka/auth/AivenSaslScramServerCallbackHandler.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.kafka.common.security.scram.ScramLoginModule;
+import org.apache.kafka.common.security.scram.internals.ScramFormatter;
+import org.apache.kafka.common.security.scram.internals.ScramMechanism;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class AivenSaslScramServerCallbackHandler implements AuthenticateCallbackHandler {
+  private static final Logger logger =
+      LoggerFactory.getLogger(AivenSaslScramServerCallbackHandler.class);
+  private String configFileLocation;
+  private ScramFormatter formatter;
+
+  @Override
+  public void configure(Map<String, ?> configs, String mechanism,
+          List<AppConfigurationEntry> jaasConfigEntries) {
+    configFileLocation = JaasContext.configEntryOption(
+            jaasConfigEntries, "users.config", ScramLoginModule.class.getName());
+    logger.info("Using configuration file {}", configFileLocation);
+    try {
+      formatter = new ScramFormatter(ScramMechanism.forMechanismName(mechanism));
+    } catch (NoSuchAlgorithmException e) {
+      logger.error("Error configuring SASL/SCRAM callback, unrecognized mechanism {}", mechanism);
+    }
+  }
+
+  @Override
+  public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+    String username = null;
+    for (Callback callback : callbacks) {
+      if (callback instanceof NameCallback) {
+        NameCallback nameCallback = (NameCallback) callback;
+        username = nameCallback.getDefaultName();
+      } else if (callback instanceof ScramCredentialCallback) {
+        ScramCredentialCallback credCallback = (ScramCredentialCallback) callback;
+        ScramCredential creds = getScramCreds(username);
+        credCallback.scramCredential(creds);
+      } else {
+        throw new UnsupportedCallbackException(callback);
+      }
+    }
+  }
+
+  /** Generate SCRAM creds for given username. */
+  public ScramCredential getScramCreds(String username) {
+    File configFile = new File(configFileLocation);
+
+    if (formatter == null) {
+      logger.error("Authentication failed for {}, no credential formatter set", username);
+      return null;
+    }
+
+    JSONParser parser = new JSONParser();
+    try {
+      Object obj = parser.parse(new FileReader(configFile));
+      JSONArray root = (JSONArray) obj;
+      Iterator<JSONObject> iter = root.iterator();
+      while (iter.hasNext()) {
+        JSONObject node = iter.next();
+        if (username.equals(node.get("username"))) {
+          String storedPassword = (String)node.get("password");
+          if (storedPassword == null) {
+            logger.error("Authentication failed for {}, no password set", username);
+            return null;
+          }
+          return formatter.generateCredential(storedPassword, 4096);
```

Comment:
This 4096 looks magical. A constant with a comment (e.g. mentioning that 4096 is the minimum allowed number of iterations) would be nice.

Reply:
Changed to use number derived from the scram method definition.

Reply:
Tests still use 4096, but with a comment